planner/cascades/transformation_rules.go

```diff
@@ -2222,6 +2225,102 @@ func (r *InjectProjectionBelowTopN) OnTransform(old *memo.ExprIter) (newExprs []
 	return []*memo.GroupExpr{topProjGroupExpr}, true, false, nil
 }
 
+// InjectProjectionBelowSort injects two Projections below and upon Sort if Sort's ByItems
+// contain ScalarFunctions.
+type InjectProjectionBelowSort struct {
+	baseRule
+}
+
+// NewRuleInjectProjectionBelowSort creates a new Transformation NewRuleInjectProjectionBelowSort.
+// The behavior and reason of this rule is almost the same as `InjectProjectionBelowTopN` above.
+// It will extract the ScalarFunctions of `ByItems` into a Projection and injects the Projection
+// below Sort. Then, we need to add another Projection upon Sort to prune the extra Columns.
+// The reason why we need this rule is that, SortExecutor in TiDB does not support ScalarFunction
+// in `ByItems`. So we have to use a Projection to calculate the ScalarFunctions in advance.
+// The pattern of this rule is: a single Sort
+func NewRuleInjectProjectionBelowSort() Transformation {
+	rule := &InjectProjectionBelowSort{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandSort,
+		memo.EngineTiDBOnly,
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *InjectProjectionBelowSort) Match(expr *memo.ExprIter) bool {
+	sort := expr.GetExpr().ExprNode.(*plannercore.LogicalSort)
+	for _, item := range sort.ByItems {
+		if _, ok := item.Expr.(*expression.ScalarFunction); ok {
+			return true
+		}
+	}
+	return false
+}
+
+// OnTransform implements Transformation interface.
+// It will convert `Sort -> X` to `Projection -> Sort -> Projection -> X`.
+func (r *InjectProjectionBelowSort) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
```

Comment:
Could we reuse the codes in `InjectProjectionBelowTopN`? I think they have the same logic.

Reply:
Seems like we can't use only one rule for two different patterns. So I tried to use a new function `generateProj4Inject` to handle some of the same logic. Is this OK?
And sorry for being late.

Reply:
Yeah, I'm OK with this solution.